contrib/signet/generate.py

```diff
@@ -0,0 +1,420 @@
+def do_solvepsbt(args):
+    block, signet_solution = do_decode_psbt(sys.stdin.read())
+    block = finish_block(block, signet_solution, args.grind_cmd)
+    print(ToHex(block))
+
+def do_generate(args):
+    if args.N != int(args.N) or args.N < -1:
+       logging.error("N must be an integer, and at least -1")
+       return 1
+
+    if args.target_mining_time is not None:
+        if args.target_mining_time <= 0 or args.target_mining_time > 600:
+            logging.error("Target mining time must be between 1 and 600")
+            return 1
+
+    mi = json.loads(args.bcli("getmininginfo"))
+    nextblock = mi["blocks"] + 1
+    mined_blocks = 0
+    last_mine_time = 0
+
+    TARGET = 600.0*2016/2015 # 10 minutes, adjusted for the off-by-one bug
+
+    if args.backdate:
+        start = args.backdate
+        logging.debug("Setting start time to %d (-backdate=%d)", start, args.backdate)
+    else:
+        start = int(time.time()) - 1
+        start -= start % args.block_time
+        start += args.block_time
+
+    while args.N <= 0 or mined_blocks < args.N:
+        # sleep
+        if args.target_mining_time:
+            block_time = TARGET * (last_mine_time / args.target_mining_time)
+            block_time = max(1, min(int(block_time), 3600)) # don't be too fast or too slow
+        else:
+            block_time = args.block_time
+
+        if block_time > 0:
+            next_time = start + block_time
+            sleep_for = next_time - time.time()
+            if sleep_for > 0:
+                time.sleep(sleep_for)
+            start = next_time
+
+        # gbt
+        tmpl = json.loads(args.bcli("getblocktemplate", '{"rules":["signet","segwit"]}'))
+        logging.debug("GBT template: %s", tmpl)
+
+        if tmpl["mintime"] > start:
+            logging.info("Updating time from %d to %d", start, tmpl["mintime"])
+            now = time.time()
+            start = tmpl["mintime"]
+            if start > now:
+                logging.error("GBT mintime is in the future: %d is %d seconds later than %d", start, (start-now), now)
+                return 1
+
+        # work out if we should actually mine
+        if args.secondary and tmpl["height"] > nextblock:
+            if nextblock is not None:
+                logging.info("Chain height increased (%d to %d), waiting", nextblock, tmpl["height"]-1)
+            nextblock = tmpl["height"]
+            continue
+        else:
+            nextblock = tmpl["height"] + 1
+
+        # address for reward
+        reward_addr, reward_spk = get_reward_addr_spk(args, tmpl["height"])
+
+        # mine block
+        logging.debug("Mining block start=%s", start)
+        mining_start = time.time()
+        mined_blocks += 1
+        psbt = generate_psbt(tmpl, reward_spk, blocktime=start)
+        psbt_signed = json.loads(args.bcli("-stdin", "walletprocesspsbt", input=psbt.encode('utf8')))
+        if not psbt_signed.get("complete",False):
+            sys.stderr.write("PSBT signing failed")
```

Comment:
Dumping the PSBT result is useful here for debugging.

Reply:
Should dump `psbt` now if debug logging is enabled. That will contain the whole block as well, so it will be long.